### PR TITLE
rimage: Allow blocks to be loaded into specific memory areas

### PR DESCRIPTION
**Pull request by dbaluta**
This will allow firmware binaries to have sections that are
loaded in different bars.

Memory areas are "manually" specified inside struct adsp. We
change that with an array of memory zones indexed by BAR type.

The indexes are specified via snd_sof_fw_blk_type which now
encodes the BAR type. To keep backward compatibility we map:
	-> SOF_BLK_TEXT to SOF_BAR_TYPE_IRAM
	-> SOF_BLK_DATA to SOF_BAR_TYPE_DRAM.

We get rid of SOF_BLK_{CACHE|REGS|SIGS} because they are not
used in SOF firmware (and ignored on Linux).

A minor drawback of this patch is that each machine description
will have memory allocated for all existing BARs even if less
are used. This can be refined in the future, by dynamically
allocating the memory zones.

Suggested-by: Liam Girdwood <liam.r.girdwood@linux.intel.com>
Signed-off-by: Daniel Baluta <daniel.baluta@nxp.com>

### CODE REVIEW

**Comment by dbaluta** [comment]
This is v2 of patch RFC https://github.com/thesofproject/sof/pull/879

**Comment by dbaluta** [comment]
@lgirdwood on the Linux side there is one more BAR (i.e XXX_PCI_BAR). I think we should also include it here, although the FW DSP might never send it. 

In this way, we could change:

 void __iomem *bar[SND_SOF_BARS];

to hold all available BARs from FW side + XXX_PCI_BAR handled only on the Linux side.

**Comment by dbaluta** [comment]
On a second thought I think it is OK to keep the list of BARs as it is on FW DSP side. And add the rest of the bars on the Linux side. 

That's because we have more bars:

#define HDA_DSP_HDA_BAR»»       »       »       0                                                                                                                                                                  
#define HDA_DSP_PP_BAR» »       »       »       1
#define HDA_DSP_SPIB_BAR»       »       »       2
#define HDA_DSP_DRSM_BAR»       »       »       3
#define HDA_DSP_BAR»    »       »       »       4

**Comment by dbaluta** [comment]
Sent v2 wheere I renamed SOF_BAR_TYPE_IRAM -> SOF_FW_BAR_TYPE_IRAM, because I will use SOF_BAR_ prefix for BARs specific to Linux side.

**Comment by dbaluta** [comment]
Yes, the naming might be a little misleading. Although, let me try to explain it better.

FW binary that is loaded by Linux kernel is created using rimage as a list of blocks. Each block contains a section from the original object file with additional information about where to load that block in memory (type, size, offset from the base).

In the current implementation "base" is the same for ALL blocks. This works fine until now because
the FW binary is really loaded in a single BAR (which is XXX_DSP_BAR).

On, i.MX (and also I think on Intel could be this requirement) we load TEXT section in local RAM
(named Instruction RAM or IRAM) and DATA section in System RAM (named SRAM).

These are two different BARs. rimage doesn't send any information about base address register.

So, when creating the FW binary image to be loaded we need to find a way to specify what is the base address for each of the sections.

The solution found with @lgirdwood in order to not break the ABI is to encode the BAR index in snd_sof_fw_blk_type of block's header.



> "I don't understand how we can map memory types used by firmware to BAR access."

Basically we create a convention between rimage and Linux kernel to map section types to BARs indexes.  I think I should really replace memory area by BAR in the commit message.

**Comment by plbossart** [comment]
On 1/29/19 2:10 PM, Daniel Baluta wrote:
>
> Yes, the naming might be a little misleading. Although, let me try to 
> explain it better.
>
> FW binary that is loaded by Linux kernel is created using rimage as a 
> list of blocks. Each block contains a section from the original object 
> file with additional information about where to load that block in 
> memory (type, size, offset from the base).
>
> In the current implementation "base" is the same for ALL blocks. This 
> works fine until now because
> the FW binary is really loaded in a single BAR (which is XXX_DSP_BAR).
>
> On, i.MX (and also I think on Intel could be this requirement) we load 
> TEXT section in local RAM
> (named Instruction RAM or IRAM) and DATA section in System RAM (named 
> SRAM).
>
> These are two different BARs. rimage doesn't send any information 
> about base address register.
>
> So, when creating the FW binary image to be loaded we need to find a 
> way to specify what is the base address for each of the sections.
>
> The solution found with @lgirdwood <https://github.com/lgirdwood> in 
> order to not break the ABI is to encode the BAR index in 
> snd_sof_fw_blk_type of block's header.
>
>     "I don't understand how we can map memory types used by firmware
>     to BAR access."
>
> Basically we create a convention between rimage and Linux kernel to 
> map section types to BARs indexes. I think I should really replace 
> memory area by BAR in the commit message.
>
My objection is on logic terms only.

On some platforms you can initialize a memory type through a BAR

On other platforms, you cannot initialize a memory type through a BAR.

Viewing memory type as a synonym for BAR isn't right. It's a hack that's 
already wrong on Intel HDaudio platforms. The loader should know which 
memory can be initialized using what BAR, we shouldn't encode the logic 
in memory types.

**Comment by dbaluta** [comment]
> My objection is on logic terms only. On some platforms you can initialize a memory type through a BAR 

> On other platforms, you cannot initialize a memory type through a BAR. 

I get this.

> Viewing memory type as a synonym for BAR isn't right. It's a hack that's already wrong on Intel HDaudio platforms. 

Ok. Makes sense. My approach seemed to fit nicely until I've met sound/soc/sof/intel/hda.h header
where I got lost at HDA_DSP_PP_BAR /  HDA_DSP_PP_BAR / HDA_DSP_PP_BAR. Not sure what are these :)

> The loader should know which memory can be initialized using what BAR, we shouldn't encode the logic in memory types.

Yes. But how? I'm now looking at snd_sof_parse_module_memcpy in sound/soc/sof/loader.c

Here the loader knows about a data block that needs to be loaded into memory the following info:
* type
* size
* offset

Let's assume that type is only SOF_BLK_DATA or SOF_BLK_TEXT 

For Baytrail it is simple because both DATA and TEXT are stored in the same continuous memory area (that starts at sdev->bar[sdev->mmio_bar).

But for others DATA and TEXT might not be in the same continous memory area. WE could abuse the "offset" field but in this case offset should be signed :)

**Comment by lgirdwood** [comment]
@dbaluta @plbossart renaming BAR to TYPE is good and I cant see any issues here since TYPE can be used to describe any memory region. It can also include platform specific and generic types.

Fwiw, the image parser for each platform can also be overridden in the driver and rimage if needed  (i.e. CNL uses manifest parser and BYT uses block parser - new parsers formats can be added too). The intention here was to support any FW image file format in tools and drivers.

**Comment by dbaluta** [comment]
> Fwiw, the image parser for each platform can also be overridden in the driver and rimage if needed (i.e. CNL uses manifest parser and BYT uses block parser - new parsers formats can be added too). The intention here was to support any FW image file format in tools and drivers.

Good thing to know that the image block parser can be overridden. With that in mind I can write my own parser for i.MX. 

Anyhow, I still think that adding information about the 'type' of memory where a block will be loaded is important! 

So, we can merge this patch together with patch 1 from PR https://github.com/thesofproject/linux/pull/604  and then let individual users to choose their implementation of the parser function.

**Comment by plbossart** [comment]
On 1/30/19 9:18 AM, Daniel Baluta wrote:
>
>     Fwiw, the image parser for each platform can also be overridden in
>     the driver and rimage if needed (i.e. CNL uses manifest parser and
>     BYT uses block parser - new parsers formats can be added too). The
>     intention here was to support any FW image file format in tools
>     and drivers.
>
> Good thing to know that the image block parser can be overridden. With 
> that in mind I can write my own parser for i.MX.
>
> Anyhow, I still think that adding information about the 'type' of 
> memory where a block will be loaded is important!
>
> So, we can merge this patch together with patch 1 from PR 
> thesofproject/linux#604 
> <https://github.com/thesofproject/linux/pull/604> and then let 
> individual users to choose their implementation of the parser function.
>
if we remove the notion of BAR I have no issue merging the addtional 
block types, e.g.

s/SOF_FW_BAR_TYPE_IRAM/SOF_FW_BLOCK_TYPE_IRAM/

**Comment by dbaluta** [comment]
> if we remove the notion of BAR I have no issue merging the addtional block types, e.g. s/SOF_FW_BAR_TYPE_IRAM/SOF_FW_BLOCK_TYPE_IRAM/

So, basically are you ok with this diff:

https://gist.github.com/dbaluta/89f8b9244a7a61bda31845ba2a5b2500

Notice that we change the name of all existing block types, but we for backward compatibility SOF_BLK_DATA is identical with SOF_BLK_TYPE_DRAM, SOF_BLK_TEXT is mapped to SOF_BLK_TYPE_IRAM.

Also, we use the new types as indexes like here:

https://gist.github.com/dbaluta/dc92d62c8e409086816212d50bd61ced

Sorry, I don't know how to paste a block of code in github comments. Using "insert code" button formats the code pretty badly.

**Comment by plbossart** [comment]
On 1/30/19 10:24 AM, Daniel Baluta wrote:
>
>     if we remove the notion of BAR I have no issue merging the
>     addtional block types, e.g.
>     s/SOF_FW_BAR_TYPE_IRAM/SOF_FW_BLOCK_TYPE_IRAM/
>
> So, basically are you ok with this diff:
>
> https://gist.github.com/dbaluta/89f8b9244a7a61bda31845ba2a5b2500
>
> Notice that we change the name of all existing block types, but we for 
> backward compatibility SOF_BLK_DATA is identical with 
> SOF_BLK_TYPE_DRAM, SOF_BLK_TEXT is mapped to SOF_BLK_TYPE_IRAM.
>
yes it's fine (you have 2 enums which are still old style).

One question though: do we need the SOF_FW_BLK_TYPE_NUM, that seems like 
a good recipe to create compatibility issues with a newer kernel?

> Also, we use the new types as indexes like here:
>
> https://gist.github.com/dbaluta/dc92d62c8e409086816212d50bd61ced
>
> Sorry, I don't know how to paste a block of code in github comments. 
> Using "insert code" button formats the code pretty badly.
>
very easy, just use ````c and ```` around the code block (doesn't work 
over email though).

**Comment by dbaluta** [comment]
> yes it's fine (you have 2 enums which are still old style). One question though: do we need the SOF_FW_BLK_TYPE_NUM, that seems like a good recipe to create compatibility issues with a newer kernel?

Yes, we need that. Because I replaced the is_dram/is_sram/is_iram family of functions with a single function get_mem_zone_type that given a section (start_addr, size) returns where in the defined memory zones it would fit, thus determining its block type.

Indeed when this number will change the ABI minor version should be bumped. 6 block types ought to be enough for everyone :D (also notice that we have one reserved value for future use, value with number 0).


> Also, we use the new types as indexes like here: https://gist.github.com/dbaluta/dc92d62c8e409086816212d50bd61ced Sorry, I don't know how to paste a block of code in github comments. Using "insert code" button formats the code pretty badly.
> very easy, just use ````c and ```` around the code block (doesn't work over email though).

I was using only one ` :))

**Comment by plbossart** [comment]
On 1/30/19 10:45 AM, Daniel Baluta wrote:
>
>     yes it's fine (you have 2 enums which are still old style). One
>     question though: do we need the SOF_FW_BLK_TYPE_NUM, that seems
>     like a good recipe to create compatibility issues with a newer kernel?
>
> Yes, we need that. Because I replaced the is_dram/is_sram/is_iram 
> family of functions with a single function get_mem_zone_type that 
> given a section (start_addr, size) returns where in the defined memory 
> zones it would fit, thus determining its block type.
>
> Indeed when this number will change the ABI minor version should be 
> bumped. 6 block types ought to be enough for everyone :D (also notice 
> that we have one reserved value for future use, value with number 0).
>
I would err on the side of adding more reserved ones, just in case. It's 
not uncommon to have different memory banks of the same type with 
different power profiles, so before you know it you could have more 
requests for block types. In general using only half of the definitions 
is recommended in standards/multiplatform solutions.

**Comment by dbaluta** [comment]
> On 1/30/19 10:45 AM, Daniel Baluta wrote: yes it's fine (you have 2 enums which are still old style). One question though: do we need the SOF_FW_BLK_TYPE_NUM, that seems like a good recipe to create compatibility issues with a newer kernel? Yes, we need that. Because I replaced the is_dram/is_sram/is_iram family of functions with a single function get_mem_zone_type that given a section (start_addr, size) returns where in the defined memory zones it would fit, thus determining its block type. Indeed when this number will change the ABI minor version should be bumped. 6 block types ought to be enough for everyone :D (also notice that we have one reserved value for future use, value with number 0).
> I would err on the side of adding more reserved ones, just in case. It's not uncommon to have different memory banks of the same type with different power profiles, so before you know it you could have more requests for block types. In general using only half of the definitions is recommended in standards/multiplatform solutions.

Agree. Ok, thanks for the comments! Will send next version asap.

**Comment by dbaluta** [comment]
@plbossart sent new version after your comments. Will follow up with change for the kernel although they independent.

**Comment by dbaluta** [comment]
@lgirdwood care to have a look?  PR looks to be already approved and also the corresponding Linux kernel change is merged.
